---
title: "Integration of single-nucleus and spatial transcriptomics reveals the molecular landscape of the human hippocampus"
output:
  github_document:
    html_preview: true
  html_document:
    toc: true
    toc_flot: true
    includes:
      in_header: header.html
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r 'setup', include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.path = "img/",
    out.width = "100%"
)
```

## Overview
Welcome to the `spatial_HPC` project! In this study, we generated spatially-resolved transcriptomics (SRT) and single-nucleus RNA-sequencing (snRNA-seq) data from adjacent tissue sections of the anterior human hippocampus across ten adult neurotypical donors. SRT data was generated using [10x
Genomics
**Visium**](https://www.10xgenomics.com/products/spatial-gene-expression) (n=36 capture areas) and [10x Genomics **Visium Spatial
Proteogenomics**
(SPG)](https://www.10xgenomics.com/products/spatial-proteogenomics) (n=8 capture areas). snRNA-seq data was generated using [10x
Genomics
**Chromium**](https://www.10xgenomics.com/products/single-cell-gene-expression) (n=26 total snRNA-seq libraries).

If you tweet about this website, the data or the R package please use
the <code>\#spatial_HPC</code> hashtag. You can find previous tweets
that way as shown
<a href="https://twitter.com/search?q=%23spatialDLPFC&src=typed_query">here</a>.

Thank you for your interest in our work!

## Study design

<img src="https://research.libd.org/spatial_hpc/img/Copy%20of%20HPC%20figure%201.png" width="1000px" align="left" />

Experimental design to generate paired single-nucleus RNA-sequencing (snRNA-seq) and spatially-resolved transcriptomics (SRT) data in the human hippocampus.
(A) Postmortem human tissue blocks from the anterior hippocampus were dissected from 10 adult neurotypical brain donors. Tissue blocks were scored and cryosectioned for snRNA-seq assays (red), and placement on Visium slides (Visium H&E, black; Visium Spatial Proteogenomics (SPG), yellow).
(B) 10Œºm tissue sections from all ten donors were placed onto 2-5 capture areas to include the extent of the HPC(n=36 total capture areas), for measurement with the 10x Genomics Visium H&E platform.
(C) 10Œºm tissue sections from two donors were placed onto 4 capture areas (n=8 total capture areas) for measurement with the 10x Genomics Visium-SPG platform.
(D) Tissue sections (2-4 100Œºm cryosections per assay) from all ten donors were collected from the same tissue blocks for  measurement with the 10x Genomics 3‚Äô gene expression platform . For each donor, we sorted on both and PI+NeuN+ (n=26 total snRNA-seq libraries).

## Interactive Websites

All of these interactive websites are powered by open source software, namely:

* üîç [`samui`](http://dx.doi.org/10.1017/S2633903X2300017X) 
* üëÄ [`iSEE`](https://doi.org/10.12688%2Ff1000research.14966.1)

We provide the following interactive websites, organized by dataset with software labeled by emojis:

* Visium (n = 44)
  - üëÄ https://libd.shinyapps.io/pseudobulk_HPC/
    - Provides tools for visualization of pseudobulked Visium data. 
  - üîç [TODO](https://samuibrowser.com/from?url=data.libd.org/samuibrowser/&s=Br3942&s=Br8325&s=Br2720&s=Br2743&s=Br3942-VSPG&s=Br6423&s=Br6432&s=Br6471&s=Br6522&s=Br8325-VSPG&s=Br8492&s=Br8667)
* snRNA-seq (n = 26)
  - üëÄ https://libd.shinyapps.io/HPC_snRNAseq_data/
    - Provides tools for visualization of snRNA-seq data.
    
## Data Access

All data, including raw FASTQ files and `SpaceRanger` processed data outputs, can be accessed via Gene Expression Omnibus (GEO) under accession XXXXXXXXXX. <cite other data access methods> 

## Contact

We value public questions, as they allow other users to learn from the
answers. If you have any questions, please ask them at
[LieberInstitute/spatial_hpc/issues](https://github.com/LieberInstitute/spatial_hpc/issues)
and refrain from emailing us. Thank you again for your interest in our
work!


## Internal

* JHPCE location: `/dcs04/lieber/lcolladotor/spatialHPC_LIBD4035/spatial_hpc`

### Files:

- `code`: Scripts for running all analyses.
- `plots`: plots generated by analysis scripts.
- `processed-data`
  - `Images`: images used for running `SpaceRanger` and `VistoSeg`.
  - `spaceranger`: `SpaceRanger` output files.
- `raw-data`
  - `sample_info`: metadata about samples.

Code for running [`GraphST`](https://doi.org/10.1038/s41467-023-36796-3) clustering pipeline can be found here: https://github.com/JianingYao/SpatialHPC_graphST_multipleSample

<!-- ## Description of HPC data -->
<!-- This is a description of data files for this project. -->

<!-- - The `/spatial_hpc/raw-data/FASTQ/` contains FASTQ files for all experiments -->
<!-- 1. The `MiSeq` and `NovaSeq` folders has softlinks to the fastqs of slides `V10B01‚àí085` and `V10B01‚àí086` -->
<!-- 2. The `2022-04-12_SPag033122` folder has softlinks to the fastqs of slides `V11A20‚àí297`, `V11L05‚àí333`, `V11L05‚àí335`, `V11L05‚àí336`, `V11U08‚àí081`, `V11U08‚àí084`. -->
<!-- - The `/spatial_hpc/raw-data/images/` contains tif files captured on CS2 for all slides -->
<!-- - The `/spatial_hpc/raw-data/sample_info/` contains information of all brains, visium slides and their master excel sheets used in the study -->


This GitHub repository is organized along the [*R/Bioconductor-powered
Team Data Science* group
guidelines](https://lcolladotor.github.io/bioc_team_ds/organizing-your-work.html#.Yaf9fPHMIdk).
It follows the
[LieberInstitute/template_project](https://github.com/LieberInstitute/template_project)
structure.
